cpp/include/cudf/quantiles.hpp

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+namespace cudf {
+
+namespace experimental {
+
+/* @brief Computes a quantile of a given column.
+ *
+ * @param[in] in            Table containing columns used to compute quantiles.
+ * @param[in] quantile      Requested quantile in range [0, 1].
+ * @param[in] interpolation Interpolation strategy for quantiles lying between
+ *                          two points.
+ * @param[in] mr            Optional. The resource to use for all allocations.
+ * @param[in] stream        Stream on which to perform computations.
+ *
+ * @returns Quantiles for each column. Elements are null if columns are empty.
+ */
+
+ 
+std::vector<std::unique_ptr<scalar>>
+quantiles(table_view const& in,
```

Comment:
When `is_sorted = true`, specifying `order`/`null_order` has no effect right now. I'm thinking we may want to change the API to something along these lines:
```
// for unsorted inputs.
std::vector<std::unique_ptr<scalar>>
quantiles(table_view const& in,
          double quantile,
          interpolation interpolation = interpolation::LINEAR);


// for tables with at least one pre-sorted column
std::vector<std::unique_ptr<scalar>>
quantiles(table_view const& in,
          double quantile,
          interpolation interpolation = interpolation::LINEAR,
          std::vector<bool> is_sorted = {},
          std::vector<cudf::order> order = {},
          std::vector<cudf::null_order> null_order = {});
```

When an input is pre-sorted, and the sort order is `cudf::order::DESCENDING`, we will look up the quantile in reverse order such that lower values are still associated with a lower quantile.

```
// psuedo code
col = { 4, 3, 2, 1 }
q_val_low = quantile_presorted(col, 0.0, order::descending) // returns 1.0
q_val_high = quantile_presorted(col, 1.0, order::descending) // returns 4.0
```

My code doesn't consider this yet. What's the expected behavior?

Reply:
Why not have just `std::vector<cudf::order> value_order = {}`, and then add `cudf::order::UNORDERED` as one of the enum values?

Reply:
Probably because I didn't want to touch `cudf::order`, but that sounds great. Adding now.

Reply:
How about something like this:
```c++
struct order_info{
   bool is_ordered;
   order o;
   null_order n;
};

quantiles(..., vector<order_info>);
```

Reply:
And consume it like so?

```
auto order_infos = { { false },
                   { true, order::ASCENDING },
                   { true, order::DESCENDING, null_order::BEFORE }};

auto qs = quantiles(..., order_infos);
```